Sources/Purchasing/CachingProductsManager.swift

```diff
@@ -115,32 +115,31 @@ private extension CachingProductsManager {
         requestCache: Atomic<[Set<String>: [(Result<Set<T>, PurchasesError>) -> Void]]>,
         fetcher: (Set<String>, @escaping (Result<Set<T>, PurchasesError>) -> Void) -> Void
     ) {
-        if let cachedProducts = Self.cachedProducts(with: identifiers, productCache: productCache) {
-            Logger.debug(
-                Strings.offering.products_already_cached(
-                    identifiers: Set(cachedProducts.map { $0.productIdentifier})
-                )
-            )
+        let cachedProducts = Self.cachedProducts(with: identifiers, productCache: productCache)
+        let missingProducts = identifiers.subtracting(cachedProducts.keys)
 
-            completion(.success(cachedProducts))
+        if missingProducts.isEmpty {
+            completion(.success(Set(cachedProducts.values)))
         } else {
-            let requestInProgress = Self.save(completion, for: identifiers, requestCache: requestCache)
+            let requestInProgress = Self.save(completion, for: missingProducts, requestCache: requestCache)
             guard !requestInProgress else {
-                Logger.debug(Strings.offering.found_existing_product_request(identifiers: identifiers))
+                Logger.debug(Strings.offering.found_existing_product_request(identifiers: missingProducts))
                 return
             }
 
             Logger.debug(
-                Strings.storeKit.no_cached_products_starting_store_products_request(identifiers: identifiers)
+                Strings.storeKit.no_cached_products_starting_store_products_request(identifiers: missingProducts)
             )
 
-            fetcher(identifiers) { result in
+            fetcher(missingProducts) { result in
```

Comment:
Ahh nvm! I think this is correct after taking a closer look. I was thinking the call to `Self.save` before this added the request to the `requestCache` and we were not clearing it from there on errors but taking a closer look, this completion block gets called for errors as well and so it would get cleared when we do `getAndClearRequestCompletion`.